main/modes/mode_gamepad.c

```diff
@@ -104,25 +154,192 @@ const hid_gamepad_button_bm_t touchMap[] =
  */
 void gamepadEnterMode(display_t* disp)
 {
+     // Allocate and zero memory
+    gm = calloc(1, sizeof(gamepadMenu_t));
+
+    gm->disp = disp;
+    
+    loadFont("mm.font", &(gm->mmFont));
+
+    gm->menu = initMeleeMenu("Gamepad", &(gm->mmFont), gamepadMainMenuCb);
+
+    setGamepadMainMenu();
+}
+
+/**
+ * Exit the gamepad mode and free memory
+ */
+void gamepadExitMode(void)
+{
+    deinitMeleeMenu(gm->menu);
+    freeFont(&(gm->mmFont));
+    free(gm);
+
+    if(gamepad != NULL){
+        freeFont(&(gamepad->ibmFont));
+        free(gamepad);
+    }
+}
+
+void setGamepadMainMenu(void)
+{
+    resetMeleeMenu(gm->menu, str_gamepadTitle, gamepadMainMenuCb);
+    addRowToMeleeMenu(gm->menu, str_pc);
+    addRowToMeleeMenu(gm->menu, str_ns);
+    addRowToMeleeMenu(gm->menu, str_exit);
+
+    gm->screen = GAMEPAD_MENU;
+}
+
+void gamepadMainMenuCb(const char* opt)
+{
+    if (opt == str_pc)
+    {
+        gamepadStart(gm->disp, GAMEPAD_GENERIC);
+        gm->screen = GAMEPAD_MAIN;
+        return;
+    }
+
+    if (opt == str_ns)
+    {
+        gamepadStart(gm->disp, GAMEPAD_NS);
+        gm->screen = GAMEPAD_MAIN;
+        return;
+    }
+
+    if (opt == str_exit)
+    {
+        // Exit to main menu
+        switchToSwadgeMode(&modeMainMenu);
+        return;
+    }
+}
+
+void gamepadStart(display_t* disp, gamepadType_t type){
     // Allocate memory for this mode
     gamepad = (gamepad_t*)calloc(1, sizeof(gamepad_t));
 
     // Save a pointer to the display
     gamepad->disp = disp;
 
+    gamepad->gamepadType = type;
+
+    tusb_desc_device_t nsDescriptor = {
+        .bLength = 18U,
+        .bDescriptorType = 1,
+        .bcdUSB = 0x0200,
+        .bDeviceClass = 0x00,
+        .bDeviceSubClass = 0x00,
+        .bDeviceProtocol = 0x00,
+        .bMaxPacketSize0 = 64,
+        .idVendor = 0x0f0d,
+        .idProduct = 0x0092,
+        .bcdDevice = 0x0100,
+        .iManufacturer = 0x01,
+        .iProduct = 0x02,
+        .iSerialNumber = 0x03,
+        .bNumConfigurations = 0x01
+    };
+
+    tinyusb_config_t default_cfg = {
+    };
+
+    tinyusb_config_t tusb_cfg = {
+        .descriptor = &nsDescriptor
+    };
+
+    tinyusb_driver_install((gamepad->gamepadType==GAMEPAD_NS) ? &tusb_cfg :&default_cfg);
```

Comment:
I guess this safely replaces the driver installed in `swadge_esp32.c`?

Reply:
Yes, to temporarily change the VID and PID while Switch mode is in use. It _should_ be safe. Works so far in my testing.

Reply:
Dug into this and `tinyusb_driver_install()` calls `usb_new_phy()`. We don't have src for that, so I'm not sure what it does, but I do know there is a corresponding `usb_del_phy()` which doesn't seem to be callable from anywhere.

`tusb_run_task()` is also called when installing the driver and has this note:
```C
esp_err_t tusb_run_task(void)
{
    // This function is not garanteed to be thread safe, if invoked multiple times without calling `tusb_stop_task`, will cause memory leak
```

Feels a little fishy to me